core/lib/zksync_core/src/api_server/tx_sender/mod.rs

```diff
@@ -897,30 +893,29 @@ impl TxSender {
             )
         } as u64;
 
-        let full_gas_limit =
-            match tx_body_gas_limit.overflowing_add(additional_gas_for_pubdata + overhead) {
-                (value, false) => {
-                    if value > max_gas_limit {
-                        return Err(SubmitTxError::ExecutionReverted(
-                            "exceeds block gas limit".to_string(),
-                            vec![],
-                        ));
-                    }
-
-                    value
-                }
-                (_, true) => {
+        let full_gas_limit = match suggested_gas_limit.overflowing_add(overhead) {
+            (value, false) => {
+                if value > max_gas_limit {
                     return Err(SubmitTxError::ExecutionReverted(
                         "exceeds block gas limit".to_string(),
                         vec![],
                     ));
                 }
-            };
+
+                value
+            }
+            (_, true) => {
+                return Err(SubmitTxError::ExecutionReverted(
+                    "exceeds block gas limit".to_string(),
+                    vec![],
+                ));
+            }
+        };
 
         tracing::info!(
             initiator = ?tx.initiator_account(),
             nonce = %tx.nonce().unwrap_or(Nonce(0)),
-            "fee estimation: gas for pubdata: {}, tx body gas: {tx_body_gas_limit}, overhead gas: {overhead} \
+            "fee estimation: gas for pubdata: {}, suggested gas: {suggested_gas_limit}, overhead gas: {overhead} \
             (with params base_fee: {base_fee}, gas_per_pubdata_byte: {gas_per_pubdata_byte})",
             (tx_metrics.pubdata_published as u64) * gas_per_pubdata_byte,
         );
```

Comment:
Let's change what we log here:
- `gas for pubdata` should be `tx_metrics.pubdata_published * gas_per_pubdata_byte * estimated_fee_scale_factor`
- tx body gas (or maybe just call it computational gas?) should be `upper_bound * estimated_fee_scale_factor`
- let's print `estimated_fee_scale_factor` param as well

Reply:
> tx body gas (or maybe just call it computational gas?) should be upper_bound * estimated_fee_scale_factor

@perekopskiy the problem with this approach is that the `additional_gas_for_pubdata` may be higher than the "correct" published gas, since it is checked under the condition of `max_gas_limit` to avoid doing binary search over gas that can be spent on pubdata only (the large gas limit means larger balance change). This is a very real scenario which I observed locally when testing large gas prices when the `upper_bound` was near zero because of `additional_gas_for_pubdata` being large enough

-

Reply:
then maybe `suggested_gas_limit - tx_metrics.pubdata_published * gas_per_pubdata_byte * estimated_fee_scale_factor`?